test/crawlers/browser_crawler.test.js

```diff
@@ -0,0 +1,507 @@
+            handleFailedRequestFunction: async ({ request }) => {
+                failedRequests.push(request);
+            },
+        });
+
+        await crawler.run();
+
+        expect(failedRequests.length).toBe(1);
+        failedRequests.forEach((fr) => {
+            const [msg] = fr.errorMessages;
+            expect(msg).toContain(`gotoFunction timed out after ${gotoTimeoutSecs} seconds.`);
+        });
+        expect(called).toBe(false);
+    });
+
+    test('should throw on "blocked" status codes (retire session)', async () => {
+        const baseUrl = 'https://example.com/';
+        const sources = STATUS_CODES_BLOCKED.map((statusCode) => {
+            return {
+                url: baseUrl + statusCode,
+                userData: { statusCode },
+            };
+        });
+        const requestList = await Apify.openRequestList(null, sources);
+
+        let called = false;
+        const failedRequests = [];
+        const crawler = new Apify.BrowserCrawler({
+            browserPlugins: [puppeteerPlugin],
+            requestList,
+            useSessionPool: true,
+            persistCookiesPerSession: false,
+            maxRequestRetries: 0,
+            gotoFunction: async ({ request }) => {
+                return { status: () => request.userData.statusCode };
+            },
+            handlePageFunction: async () => { // eslint-disable-line no-loop-func
+                called = true;
+            },
+            handleFailedRequestFunction: async ({ request }) => {
+                failedRequests.push(request);
+            },
+        });
+
+        await crawler.run();
+
+        expect(failedRequests.length).toBe(STATUS_CODES_BLOCKED.length);
+        failedRequests.forEach((fr) => {
+            const [msg] = fr.errorMessages;
+            expect(msg).toContain(`Request blocked - received ${fr.userData.statusCode} status code.`);
+        });
+        expect(called).toBe(false);
+    });
+
+    describe('proxy', () => {
+        let requestList;
+        beforeEach(async () => {
+            requestList = new Apify.RequestList({
+                sources: [
+                    { url: 'http://example.com/?q=1' },
+                    { url: 'http://example.com/?q=2' },
+                    { url: 'http://example.com/?q=3' },
+                    { url: 'http://example.com/?q=4' },
+                ],
+            });
+            await requestList.initialize();
+        });
+
+        afterEach(() => {
+            requestList = null;
+        });
+
+        test('browser should launch with correct proxyUrl', async () => {
+            process.env[ENV_VARS.PROXY_PASSWORD] = 'abc123';
+            const status = { connected: true };
+            const fakeCall = async () => {
+                return { body: status };
+            };
+
+            const stub = sinon.stub(utilsRequest, 'requestAsBrowser').callsFake(fakeCall);
+            const proxyConfiguration = await Apify.createProxyConfiguration();
```

Comment:
You can use the `proxyUrls` or `newUrlFunction` option of `ProxyConfiguration` to avoid the `requestAsBrowser` check.